[FSDP2] better error msg for cpu offloading

weifengpy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135156

when cpu offloading is enabled, if user load a gpu state dict, FSDP2 will throw a less obvious error at backward
```
RuntimeError: attempting to assign a gradient with device type 'cpu' to a tensor with device type 'cuda'. Please ensure that the gradient and the tensor are on the same device
```

this PR throws error more explicitly by specifying which parameters should be moved because of cpu offloading

```
FSDP parameters should be materialized on cpu when enabling cpu offloading. For example, load cpu state dict or call module.to_empty(device="cpu"). Found following parameters on non-cpu device: ['0.weight']
```

`pytest -s test/distributed/_composable/fsdp/test_fully_shard_state_dict.py -k test_dp_state_dict_cpu_offload`


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

weifengpy committed:
[FSDP2] move DTensor and local tensor to cpu for cpu offloading

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

weifengpy commented on torch/distributed/_composable/fsdp/_fsdp_param.py:
```diff
@@ -716,11 +715,20 @@ def reset_sharded_param(self):
+            self.to_sharded_dtensor(local_tensor[: self.sharded_size[0]]),
+            requires_grad=new_param.requires_grad,
+        )
+        self._setattr_on_modules(self.sharded_param)
```
this won't work because optimizer.param_groups refers to a different `model.parameters()`

weifengpy committed:
Update on "[FSDP2] move DTensor and local tensor to cpu for cpu offloading"



`pytest -s distributed/_composable/fsdp/test_fully_shard_training.py -k test_to_float64_after_init`

resolve cpu offload error in TorchTune: https://github.com/pytorch/torchtune/issues/1412

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy commented on torch/distributed/_composable/fsdp/_fsdp_param.py:
```diff
@@ -707,7 +707,8 @@ def reset_sharded_param(self):
                 )
-            self.sharded_param = new_param
+        elif skip_same_param:
+            return
```
reset_sharded_param has 3 callsites
* state dict load
* _apply, eg model.to(dtype)
* lazy_init

for lazy_init, we should return early if the param stays the same. otherwise at the end we create new param with `nn.Parameter()` and that will obsolete optim's reference to `model.parameters()`

awgu replied:
Wait, so is `main` branch wrong today ignoring this PR?

FSDP's `self.sharded_param` is a different tensor than the one referenced by the optimizer?

If so, I think we should fix this bug separately.

weifengpy replied:
haha. trunk is healthy because we do `self.sharded_param._local_tensor = local_tensor[: self.sharded_size[0]]` in trunk. that won't create a new `nn.Parameter()`

weifengpy commented on torch/distributed/_composable/fsdp/_fsdp_param.py:
```diff
@@ -716,11 +717,20 @@ def reset_sharded_param(self):
             padded_local_tensor[: local_tensor.size(0)].copy_(local_tensor)
             local_tensor = padded_local_tensor
+        if self.offload_to_cpu and not local_tensor.is_cpu:
+            local_tensor = local_tensor.cpu()
```
cover `pin_memory=False`

weifengpy committed:
Update on "[FSDP2] move DTensor and local tensor to cpu for cpu offloading"



`pytest -s distributed/_composable/fsdp/test_fully_shard_training.py -k test_to_float64_after_init`

resolve cpu offload error in TorchTune: https://github.com/pytorch/torchtune/issues/1412

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy commented on torch/distributed/_composable/fsdp/_fsdp_param.py:
```diff
@@ -716,11 +717,20 @@ def reset_sharded_param(self):
+            raise ValueError(
+                f"expect the new parameter to be DTensor but got {type(new_param)}"
+            )
+        self.sharded_param = nn.Parameter(
```
making sure DTensor.device is consistent with local_tensor, which can be `cpu`

weifengpy committed:
Update on "[FSDP2] construct DTensor parameters from cpu offloaded local tensors"



`pytest -s distributed/_composable/fsdp/test_fully_shard_training.py -k test_to_float64_after_init`

resolve cpu offload error in TorchTune: https://github.com/pytorch/torchtune/issues/1412

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy commented:
synced and I will modify the PR to throw error for gpu state dict instead of moving gpu state dict to cpu implicitly

weifengpy committed:
Update on "[FSDP2] construct DTensor parameters from cpu offloaded local tensors"


resolve cpu offload error in TorchTune: https://github.com/pytorch/torchtune/issues/1412

this PR constructs DTensor from cpu offloaded local tensor

`pytest -s test/distributed/_composable/fsdp/test_fully_shard_state_dict.py -k test_dp_state_dict_cpu_offload`


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP2] construct DTensor parameters from cpu offloaded local tensors"


resolve cpu offload error in TorchTune: https://github.com/pytorch/torchtune/issues/1412

this PR constructs DTensor from cpu offloaded local tensor

`pytest -s test/distributed/_composable/fsdp/test_fully_shard_state_dict.py -k test_dp_state_dict_cpu_offload`


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP2] construct DTensor parameters from cpu offloaded local tensors"


resolve cpu offload error in TorchTune: https://github.com/pytorch/torchtune/issues/1412

this PR constructs DTensor from cpu offloaded local tensor

`pytest -s test/distributed/_composable/fsdp/test_fully_shard_state_dict.py -k test_dp_state_dict_cpu_offload`


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP2] construct DTensor parameters from cpu offloaded local tensors"


resolve cpu offload error in TorchTune: https://github.com/pytorch/torchtune/issues/1412

this PR constructs DTensor from cpu offloaded local tensor

`pytest -s test/distributed/_composable/fsdp/test_fully_shard_state_dict.py -k test_dp_state_dict_cpu_offload`


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP2] construct DTensor parameters from cpu offloaded local tensors"


resolve cpu offload error in TorchTune: https://github.com/pytorch/torchtune/issues/1412

this PR constructs DTensor from cpu offloaded local tensor

`pytest -s test/distributed/_composable/fsdp/test_fully_shard_state_dict.py -k test_dp_state_dict_cpu_offload`


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP2] better error msg for cpu offloading"


when cpu offloading is enabled, if user load a gpu state dict, FSDP2 will throw a less obvious error at backward
```
RuntimeError: attempting to assign a gradient with device type 'cpu' to a tensor with device type 'cuda'. Please ensure that the gradient and the tensor are on the same device
```

this PR throws error more explicitly by specifying which parameters should be moved because of cpu offloading

```
FSDP parameters should be materialized on cpu when enabling cpu offloading. For example, load cpu state dict or call module.to_empty(device="cpu"). Found following parameters on non-cpu device: {param_names_not_on_cpu}
```

`pytest -s test/distributed/_composable/fsdp/test_fully_shard_state_dict.py -k test_dp_state_dict_cpu_offload`


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy commented:
@awgu I repurposed the PR to throw error msg when loading gpu state dict. ready for review

awgu commented on torch/distributed/_composable/fsdp/_fsdp_param_group.py:
```diff
@@ -569,6 +560,35 @@ def _with_fqn(self, label: str) -> str:
+        ]
+        if param_names_not_on_cpu:
+            raise RuntimeError(
+                "FSDP parameters should be materialized on cpu when enabling cpu offloading. "
```
nit: I think we can capitalize CPU
```suggestion
                "FSDP parameters should be materialized on CPU when enabling CPU offloading. "
```

awgu commented on torch/distributed/_composable/fsdp/_fsdp_param_group.py:
```diff
@@ -569,6 +560,35 @@ def _with_fqn(self, label: str) -> str:
+        if param_names_not_on_cpu:
+            raise RuntimeError(
+                "FSDP parameters should be materialized on cpu when enabling cpu offloading. "
+                'For example, load cpu state dict or call module.to_empty(device="cpu"). '
```
```suggestion
                'For example, load a CPU state dict or call module.to_empty(device="cpu"). '
```

awgu commented on torch/distributed/_composable/fsdp/_fsdp_param_group.py:
```diff
@@ -569,6 +560,35 @@ def _with_fqn(self, label: str) -> str:
+    def _validate_cpu_offload_params(self):
+        if not isinstance(self.offload_policy, CPUOffloadPolicy):
+            return
+        param_names_not_on_cpu = [
+            fsdp_param._param_fqn
+            for fsdp_param in self.fsdp_params
+            if fsdp_param.sharded_param.device.type != "cpu"
+        ]
```
This is just a suggestion related to below. Specifically, I think it would be helpful to include the device in the error message in this case.


```suggestion
        fsdp_params_not_on_cpu = [
            fsdp_param
            for fsdp_param in self.fsdp_params
            if fsdp_param.sharded_param.device.type != "cpu"
        ]
```

awgu commented on torch/distributed/_composable/fsdp/_fsdp_param_group.py:
```diff
@@ -569,6 +560,35 @@ def _with_fqn(self, label: str) -> str:
+            raise RuntimeError(
+                "FSDP parameters should be materialized on cpu when enabling cpu offloading. "
+                'For example, load cpu state dict or call module.to_empty(device="cpu"). '
+                f"Found following parameters on non-cpu device: {param_names_not_on_cpu}\n"
```
part 2 of the suggestion of including the device in the error message
(needs some formatting)
```suggestion
                f"Found following parameters on non-cpu device: {[(fsdp_param._param_fqn, fsdp_param.sharded_param.device) for fsdp_param in fsdp_params_not_on_cpu]]}\n"
```

awgu reviewed (approved):
just one suggestion for including the device in the error message for `_validate_cpu_offload_params`

weifengpy committed:
Update on "[FSDP2] better error msg for cpu offloading"


when cpu offloading is enabled, if user load a gpu state dict, FSDP2 will throw a less obvious error at backward
```
RuntimeError: attempting to assign a gradient with device type 'cpu' to a tensor with device type 'cuda'. Please ensure that the gradient and the tensor are on the same device
```

this PR throws error more explicitly by specifying which parameters should be moved because of cpu offloading

```
FSDP parameters should be materialized on cpu when enabling cpu offloading. For example, load cpu state dict or call module.to_empty(device="cpu"). Found following parameters on non-cpu device: ['0.weight']
```

`pytest -s test/distributed/_composable/fsdp/test_fully_shard_state_dict.py -k test_dp_state_dict_cpu_offload`


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy commented:
@pytorchmergebot merge

weifengpy commented:
@pytorchmergebot merge

weifengpy committed:
Update on "[FSDP2] better error msg for cpu offloading"


when cpu offloading is enabled, if user load a gpu state dict, FSDP2 will throw a less obvious error at backward
```
RuntimeError: attempting to assign a gradient with device type 'cpu' to a tensor with device type 'cuda'. Please ensure that the gradient and the tensor are on the same device
```

this PR throws error more explicitly by specifying which parameters should be moved because of cpu offloading

```
FSDP parameters should be materialized on cpu when enabling cpu offloading. For example, load cpu state dict or call module.to_empty(device="cpu"). Found following parameters on non-cpu device: ['0.weight']
```

`pytest -s test/distributed/_composable/fsdp/test_fully_shard_state_dict.py -k test_dp_state_dict_cpu_offload`


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

weifengpy commented:
@pytorchmergebot merge
